### DISCUSSION TITLE
Generating threads inside class instance stops execution

### DISCUSSION THREAD

**Discussion by lemmuh** [mruby/mruby#6077] [Q&A]
I'm using the mruby-thread gem in order to use threads in my app.

Solution 1

```
100.times.each_with_object([]) do |i, a|
  a << Thread.new do
    p i
    Thread.sleep 1
  end
end.each { |t| t.join }
```

works while solution 2 

```
class A
  attr_reader :threads

  def initialize
    @threads = 100.times.each_with_object([]) do |i, a|
      a << Thread.new do
        p i
        Thread.sleep 1
      end
    end
  end
end

A.new.threads.each { |t| t.join }
```

is terminating the app with the error

```(unknown):0: undefined class/module A (ArgumentError)```

Does anybody know what is wrong with solution 2 ?

**Comment by dearblue** [comment]
This is because any outer `self` or variable referenced by the block passed to Fiber.new is an instance of class `A`.
As I wrote in https://github.com/mattn/mruby-thread/issues/68#issuecomment-1763291722, I think you need to create your own mrbgem or extend mruby-thread.
